File: lambdas/python/cpu-intensive/handler.py
from __future__ import annotations

import hashlib
import json
import logging
import platform
from typing import Any

logger = logging.getLogger()
logger.setLevel(logging.INFO)

DEFAULT_ITERATIONS: int = 1_000_000
MAX_ITERATIONS: int = 10_000_000


def lambda_handler(event: dict[str, Any] | None, context) -> dict[str, Any]:
    """Lambda handler - CPU intensive test executes SHA-256 hashing iterations to measure CPU performance.

    Executes repeated SHA-256 hashing in a tight loop to measure raw compute
    performance differences between architectures and runtimes.
    """
    event = event or {}

    logger.info(json.dumps({
        "event": "handler_start",
        "workloadType": "cpu-intensive",
        "runtime": f"python{platform.python_version()}",
        "architecture": platform.machine(),
        "requestId": getattr(context, "aws_request_id", "unknown")
    }))

    try:
        iterations = int(event.get("iterations", DEFAULT_ITERATIONS))
    except Exception:
        return _fail("invalid 'iterations' (must be an integer)")

    if iterations <= 0:
        return _fail("iterations must be > 0")
    if iterations > MAX_ITERATIONS:
        return _fail(f"iterations too high (max {MAX_ITERATIONS})")

    try:
        result_hex = _cpu_sha256(iterations)

        logger.info(json.dumps({
            "event": "handler_success",
            "iterations": iterations,
            "resultHashLength": len(result_hex)
        }))

        return {
            "success": True,
            "workloadType": "cpu-intensive",
            "iterations": iterations,
            "architecture": platform.machine(),
            "pythonVersion": platform.python_version(),
            "memoryLimitMB": int(getattr(context, "memory_limit_in_mb", 0) or 0),
            "resultHash": result_hex,  # 64-char hex
        }
    except Exception as e:
        logger.error(json.dumps({
            "event": "handler_error",
            "errorType": type(e).__name__,
            "errorMessage": str(e)
        }))
        return _fail(f"{type(e).__name__}: {e}")


def _cpu_sha256(iterations: int) -> str:
    """Chains SHA-256 hashes together for CPU stress testing."""
    sha256 = hashlib.sha256
    data = b"benchmark data for Lambda ARM vs x86 performance testing"
    for _ in range(iterations):
        data = sha256(data).digest()
    return data.hex()


def _fail(msg: str) -> dict[str, Any]:
    """Return error response in standard format."""
    return {"success": False, "workloadType": "cpu-intensive", "error": msg}

File: lambdas/nodejs/cpu-intensive/handler.ts
import type { Context } from "aws-lambda";
import { createHash } from "crypto";

const DEFAULT_ITERATIONS = 1_000_000;
const MAX_ITERATIONS = 10_000_000;

interface BenchmarkEvent {
  iterations?: number;
}

interface BenchmarkSuccess {
  success: true;
  workloadType: 'cpu-intensive';
  iterations: number;
  architecture: string;
  memoryLimitMB: number;
  resultHash: string;
}

interface BenchmarkError {
  success: false;
  workloadType: 'cpu-intensive';
  error: string;
}

type BenchmarkResult = BenchmarkSuccess | BenchmarkError;

/**
 * Lambda handler - CPU intensive test executes SHA-256 hashing iterations to measure CPU performance
 */
export async function handler(event: BenchmarkEvent, context: Context): Promise<BenchmarkResult> {
  console.log(
    JSON.stringify({
      event: 'handler_start',
      workloadType: 'cpu-intensive',
      runtime: `nodejs${process.version}`,
      architecture: process.arch,
      requestId: context.awsRequestId
    })
  );

  const iterations = event.iterations ?? DEFAULT_ITERATIONS;

  if (iterations > MAX_ITERATIONS) {
    return {
      success: false,
      workloadType: 'cpu-intensive',
      error: `iterations too high (max ${MAX_ITERATIONS})`
    };
  }

  try {
    const result = cpuIntensiveWorkload(iterations);

    console.log(
      JSON.stringify({
        event: 'handler_success',
        iterations,
        resultHashLength: result.length
      })
    );

    return {
      success: true,
      workloadType: 'cpu-intensive',
      iterations,
      architecture: process.arch,
      memoryLimitMB: parseInt(context.memoryLimitInMB, 10),
      resultHash: result.slice(0, 64)
    };
  } catch (error) {
    console.error(
      JSON.stringify({
        event: 'handler_error',
        errorType: error instanceof Error ? error.constructor.name : 'Unknown',
        errorMessage: error instanceof Error ? error.message : String(error)
      })
    );

    return {
      success: false,
      workloadType: 'cpu-intensive',
      error: error instanceof Error ? error.message : String(error)
    };
  }
}

/**
 * Chains SHA-256 hashes together for CPU stress testing
 */
function cpuIntensiveWorkload(iterations: number): string {
  let data = Buffer.from('benchmark data for Lambda ARM vs x86 performance testing');

  for (let i = 0; i < iterations; i++) {
    data = createHash('sha256').update(data).digest(); // Keep as Buffer (more efficient)
  }

  return data.toString('hex');
}